### ISSUE TITLE
"ls | grep" is needed with -A/-B

### ISSUE THREAD

**Issue by Fabian42** [koalaman/shellcheck#2389]
#### For bugs
- Rule ID: SC2010
- My shellcheck version: 0.7.2
- [x] The rule's wiki page does not already cover this
- [x] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [ ] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash
ls | grep -A9999 "^$1\$"
```

#### Here's what shellcheck currently says:
`^-- SC2010: Don't use ls | grep. Use a glob or a for loop with a condition to allow non-alphanumeric filenames.`

#### Here's what I wanted or expected to see:
Nothing. `--after-context`/`-A`, `--before-context`/`-B` cannot be replaced with globs. Or maybe, if it can, the code would likely be much more confusing. A `for` loop would definitely not make this any better either.
The example is used in [my `alldiff` function](https://github.com/Fabian42/bash_scripts/blob/c74f916afafaf0973546e3102702f2bf18e7a1cc/.bashrc#L236) to loop over all pairs of files without checking each pair twice, like this: AB, AC, AD, BC, BD, CD)

**Comment by Dmole** [comment]
Firstly 

    ls -1

is more likely to be what you want.
Secondly grep supports non-alphanumeric eg;

    grep -P "[\w\W]"

Thirdly shellcheck often wants null terminated lines like this;

    find . -maxdepth 1 -print0 | grep -Z -A9999 "^./$1\$"

so it looks like shellcheck is showing the wrong error vs a false positive.

**Comment by Fabian42** [comment]
I do actually have `ls -1` in the actual function, I just missed that when writing the minimal example here.
Even if `ls | grep ` was in some way discouraged usually, from all I know, it is actually required in this case.

**Comment by brother** [comment]
If you have determined the suggestion as not relevant for your case you can add an ignore directive to the line and continue to use it. Not sure if relevant to check if one of the flags to grep is making it a more valid use case than just offloading it to the user and teaching about a general usage way.

**Comment by Dmole** [comment]
MacOS grep is limited (no -P option) but anything GNU/Linux can handle any regex;

grep -F :
grep    : ^$[]*.\
grep -E : ()|+?{}. \1
grep -P : \w \s \d \b *? +? (?

**Comment by koalaman** [comment]
ShellCheck is right in that this construct is causing the function to fail for non-alphanumeric filenames, and that using a for loop instead would fix it. For example, if you `touch file{' 1',' 2',3}.txt` then it will output

```
file 1.txt\nfile3.txt\n
file 2.txt\nfile3.txt\n
```

but fail to realize that `file 1.txt` is also the same as `file 2.txt`.

If you replaced `ls | grep` with a condition in a for loop as suggested, e.g.

```
alldiff(){
 for file1 in *; do
  found=0
  for file2 in *; do
    if [[ $found = 0 ]]
    then
      [[ $file1 = $file2 ]] && found=1
      continue
    fi

    diff -- "$file1" "$file2" &> /dev/null
    if [[ $? == 0 ]]; then
      echo "$file1\n$file2\n"
    fi
  done
 done
}
```

then it would work for all filenames (though an array with two arithmetic for loops would be even better)

You can definitely choose to make the call that this extra complexity is not worth the improved correctness given your system's filenames (in which case you can ignore it with a directive), but since the suggestion is valid and useful, it does not make sense to suppress it generally.

**Comment by superbonaci** [comment]
@koalaman don't you must use quotation marks also in the second comparison:
```
      [[ "$file1" = "$file2" ]] && found=1
```
